source/extensions/filters/http/response_map/response_map_filter.cc

```diff
@@ -0,0 +1,254 @@
+  }
+
+  // If we decided not to rewrite the response, simply pass through to other
+  // filters.
+  if (!do_rewrite_) {
+    return Http::FilterDataStatus::Continue;
+  }
+
+  // We decided to rewrite the response, so drain any data received from the
+  // upstream since we're rewriting it anyway. This both prevents unnecessary
+  // buffering and also prevents generating errors if the response is too big
+  // to be buffered completely.
+  data.drain(data.length());
+
+  // The stream is not yet complete, and we can't let other filters run
+  // until we've rewritten the response body.
+  if (!end_stream) {
+    return Http::FilterDataStatus::StopIterationAndBuffer;
+  }
+
+  // Now that the stream is complete, rewrite the response body.
+  ENVOY_LOG(trace, "response map filter: encodeData doing rewrite");
+  doRewrite();
+  return Http::FilterDataStatus::Continue;
+}
+
+void ResponseMapFilter::doRewrite(void) {
+  const Buffer::Instance* encoding_buffer = encoder_callbacks_->encodingBuffer();
+
+  ENVOY_LOG(trace, "response map filter: doRewrite with {} encoding_buffer",
+            encoding_buffer != nullptr ? "non-null" : "null");
+
+  // If this route is disabled, we should never be doing a rewrite.
+  // In fact, we never should have even checked if we should do
+  // a rewrite.
+  ASSERT(!disabled_);
+
+  // We should either see no encoding buffer or an empty encoding buffer.
+  //
+  // We'll see no encoding buffer if the upstream response was never observed
+  // (ie: due to error) or if the upstream response was headers-only. Otherwise,
+  // if we did see a response, we should have drained any data we saw.
+  ASSERT(encoding_buffer == nullptr || encoding_buffer->length() == 0);
+
+  // This should be impossible, because we only set response_map_ to the
+  // global filter config (guaranteed to exist if this filter exists)
+  // or to a non-null per-route config. Guard against it anyway.
+  if (response_map_ == nullptr) {
+    ENVOY_LOG(trace,
+              "response map filter: doRewrite has no response_map_ to rewrite with, doing nothing");
+    return;
+  }
+
+  const ResponseMap::ResponseMapPtr& response_map = *response_map_;
+
+  // Fill in new_body and new_content_type using the response map rewrite.
+  // Pass in the request and response headers that we observed on the
+  // decodeHeaders and encodeHeaders paths, respectively.
+  std::string new_body;
+  absl::string_view new_content_type;
+  auto code = static_cast<Http::Code>(Http::Utility::getResponseStatus(*response_headers_));
+  response_map->rewrite(request_headers_, *response_headers_, encoder_callbacks_->streamInfo(),
+                        code, new_body, new_content_type);
+
+  // Encoding buffer may be null here even if we saw data in encodeData above. This
+  // happens when sendLocalReply sends a response downstream. By adding encoded data
+  // here, we do successfully override the sendLocalReply body, but it's not clear
+  // how or why.
```

Comment:
We should get some clarity around this condition here and rewrite the comment to explain what's going on instead of asking questions.